File: lib/rankings.ts
import { fetchHouseMembers, fetchSenateMembers, type HouseMember, type SenateMember } from "./congress"
import { fetchQuiverWithRetry, QuiverCircuitOpenError } from "./quiver"
import { getCache, setCache } from "./cache"
import { getPortfoliosFromDb, upsertPortfolios, writeBack, type DbPortfolio } from "./db"
import { persistHoldings, type HoldingPosition, type MemberHoldings } from "./stockLeaderboard"

export const HOUSE_RANKINGS_KEY = "house-rankings"
export const SENATE_RANKINGS_KEY = "senate-rankings"
export const RANKINGS_TTL_SECONDS = 2 * 60 * 60
const FANOUT_CONCURRENCY = 1
const FANOUT_DELAY_MS = 1500
const CIRCUIT_OPEN_PAUSE_MS = 65_000
const MAX_CIRCUIT_WAITS = 3

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

type RankingRow<M> = M & {
  netWorth: number | null
  stockHoldings: number | null
}
type FanoutRow<M> = RankingRow<M> & { positions: HoldingPosition[]; ok: boolean }

function stripFanoutExtras<M>(row: FanoutRow<M>): RankingRow<M> {
  const { positions, ok, ...rest } = row
  void positions
  void ok
  return rest as RankingRow<M>
}

export type RankingsPayload<M> = {
  byNetWorth: RankingRow<M>[]
  byStockHoldings: RankingRow<M>[]
  generatedAt: string
}

type QuiverTabResponse = {
  holdings_data?: {
    politician_net_worth_live?: string
  }
  live_stock_portfolio?: {
    live_stock_portfolio?: string
  }
}

const QUIVER_HEADERS = {
  "Content-Type": "application/json",
  "User-Agent": "CongressMoneyMan/1.0",
  "X-Requested-With": "XMLHttpRequest",
}

function parseJsonArray(value: string | undefined): unknown[] {
  if (!value) return []
  try {
    const parsed = JSON.parse(value) as unknown
    return Array.isArray(parsed) ? parsed : []
  } catch {
    return []
  }
}

function getLiveNetWorth(payload: QuiverTabResponse): number | null {
  const values = parseJsonArray(payload.holdings_data?.politician_net_worth_live)
  const value = values[0]
  return typeof value === "number" && Number.isFinite(value) ? value : null
}

function getLiveStockHoldings(payload: QuiverTabResponse): number | null {
  const positions = parseJsonArray(payload.live_stock_portfolio?.live_stock_portfolio)
  let total = 0
  let foundPosition = false
  for (const position of positions) {
    if (!Array.isArray(position)) continue
    const value = position[1]
    if (typeof value !== "number" || !Number.isFinite(value)) continue
    total += value
    foundPosition = true
  }
  return foundPosition ? total : null
}

function getLivePositions(payload: QuiverTabResponse): HoldingPosition[] {
  const positions = parseJsonArray(payload.live_stock_portfolio?.live_stock_portfolio)
  const out: HoldingPosition[] = []
  for (const position of positions) {
    if (!Array.isArray(position)) continue
    const value = position[1]
    if (typeof value !== "number" || !Number.isFinite(value) || value <= 0) continue
    const symbol = typeof position[0] === "string" ? position[0].trim().toUpperCase() : ""
    if (!symbol) continue
    out.push({ ticker: symbol, value })
  }
  return out
}

function compareRankingValues(left: number | null, right: number | null): number {
  if (left === null && right === null) return 0
  if (left === null) return 1
  if (right === null) return -1
  return right - left
}

export async function mapWithConcurrency<T, R>(
  items: T[],
  limit: number,
  mapper: (item: T) => Promise<R>,
  delayMs = 0
): Promise<R[]> {
  const results = new Array<R>(items.length)
  let currentIndex = 0

  async function worker() {
    while (currentIndex < items.length) {
      const nextIndex = currentIndex
      currentIndex += 1
      results[nextIndex] = await mapper(items[nextIndex])
      if (delayMs > 0 && currentIndex < items.length) await sleep(delayMs)
    }
  }

  await Promise.all(Array.from({ length: Math.min(limit, items.length) }, () => worker()))
  return results
}
const inFlight = new Map<string, Promise<unknown>>()

function singleFlight<T>(key: string, run: () => Promise<T>): Promise<T> {
  const existing = inFlight.get(key) as Promise<T> | undefined
  if (existing) return existing

  const promise = run().finally(() => inFlight.delete(key))
  inFlight.set(key, promise)
  return promise
}

async function getRankingRow<M extends { id: string; name: string }>(
  member: M
): Promise<FanoutRow<M>> {
  for (let waits = 0; ; waits++) {
    try {
      const response = await fetchQuiverWithRetry(
        `https://www.quiverquant.com/get_politician_page_tab_data/${member.id}`,
        { headers: QUIVER_HEADERS, next: { revalidate: 3600 } }
      )

      if (!response.ok) {
        console.warn(`[rankings] ${member.id} tab-data HTTP ${response.status}`)
        return { ...member, netWorth: null, stockHoldings: null, positions: [], ok: false }
      }

      const payload = (await response.json()) as QuiverTabResponse
      return {
        ...member,
        netWorth: getLiveNetWorth(payload),
        stockHoldings: getLiveStockHoldings(payload),
        positions: getLivePositions(payload),
        ok: true,
      }
    } catch (error) {
      if (error instanceof QuiverCircuitOpenError && waits < MAX_CIRCUIT_WAITS) {
        console.warn(
          `[rankings] circuit open — pausing ${CIRCUIT_OPEN_PAUSE_MS / 1000}s before retrying ${member.id}`
        )
        await sleep(CIRCUIT_OPEN_PAUSE_MS)
        continue
      }
      console.warn(
        `[rankings] ${member.id} fan-out failed:`,
        error instanceof Error ? error.message : error
      )
      return { ...member, netWorth: null, stockHoldings: null, positions: [], ok: false }
    }
  }
}

function buildPayload<M extends { name: string }>(
  rows: RankingRow<M>[]
): RankingsPayload<M> {
  const byStockHoldings = [...rows].sort((l, r) => {
    const v = compareRankingValues(l.stockHoldings, r.stockHoldings)
    return v !== 0 ? v : l.name.localeCompare(r.name)
  })

  const byNetWorth = [...rows].sort((l, r) => {
    const v = compareRankingValues(l.netWorth, r.netWorth)
    return v !== 0 ? v : l.name.localeCompare(r.name)
  })

  return { byNetWorth, byStockHoldings, generatedAt: new Date().toISOString() }
}

function countPopulated<M>(rows: RankingRow<M>[]): number {
  return rows.filter(
    (row) => row.netWorth !== null || row.stockHoldings !== null
  ).length
}
function membersToEmptyPayload<M extends { id: string; name: string }>(
  members: M[]
): RankingsPayload<M> {
  return buildPayload(
    members.map((member) => ({ ...member, netWorth: null, stockHoldings: null }))
  )
}

function mergeWithPrevious<M extends { id: string; name: string }>(
  fresh: RankingsPayload<M>,
  previous: RankingsPayload<M> | null
): RankingsPayload<M> {
  if (!previous) return fresh

  const previousById = new Map(previous.byNetWorth.map((row) => [row.id, row]))
  const mergedRows = fresh.byNetWorth.map((row) => {
    const prior = previousById.get(row.id)
    if (!prior) return row
    return {
      ...row,
      netWorth: row.netWorth ?? prior.netWorth,
      stockHoldings: row.stockHoldings ?? prior.stockHoldings,
    }
  })

  return buildPayload(mergedRows)
}

async function persistRankings<M extends { id: string; name: string }>(
  key: string,
  fresh: RankingsPayload<M>
): Promise<RankingsPayload<M>> {
  const previous = await getCache<RankingsPayload<M>>(key)
  const merged = mergeWithPrevious(fresh, previous)

  await setCache(key, merged, RANKINGS_TTL_SECONDS)
  console.log(
    `[rankings] cached ${key}: ${countPopulated(merged.byNetWorth)}/${merged.byNetWorth.length} members populated`
  )
  const portfolioRows: DbPortfolio[] = merged.byNetWorth
    .filter((row) => row.netWorth !== null || row.stockHoldings !== null)
    .map((row) => ({
      bioguide_id: row.id,
      net_worth: row.netWorth,
      stock_holdings: row.stockHoldings,
    }))
  writeBack(() => upsertPortfolios(portfolioRows))

  return merged
}

async function rankingsFromDb<M extends { id: string; name: string }>(
  members: M[]
): Promise<RankingsPayload<M> | null> {
  const portfolios = await getPortfoliosFromDb()
  if (portfolios.length === 0) return null

  const byId = new Map(portfolios.map((p) => [p.bioguide_id, p]))
  const rows = members.map((member) => {
    const portfolio = byId.get(member.id)
    return {
      ...member,
      netWorth: portfolio?.net_worth ?? null,
      stockHoldings: portfolio?.stock_holdings ?? null,
    }
  })
  return buildPayload(rows)
}

function resolveCongressApiKey(): string {
  const apiKey = process.env.CONGRESS_API_KEY ?? process.env.CONGRESS_GOV_API_KEY
  if (!apiKey) throw new Error("Missing CONGRESS_API_KEY")
  return apiKey
}

export async function computeHouseRankings(
  apiKey = resolveCongressApiKey()
): Promise<RankingsPayload<HouseMember>> {
  const members = await fetchHouseMembers(apiKey)
  const rows = await mapWithConcurrency(members, FANOUT_CONCURRENCY, getRankingRow, FANOUT_DELAY_MS)
  return buildPayload(rows.map(stripFanoutExtras))
}

export async function computeSenateRankings(
  apiKey = resolveCongressApiKey()
): Promise<RankingsPayload<SenateMember>> {
  const members = await fetchSenateMembers(apiKey)
  const rows = await mapWithConcurrency(members, FANOUT_CONCURRENCY, getRankingRow, FANOUT_DELAY_MS)
  return buildPayload(rows.map(stripFanoutExtras))
}
export async function refreshHouseRankings(
  apiKey = resolveCongressApiKey()
): Promise<RankingsPayload<HouseMember>> {
  return singleFlight(HOUSE_RANKINGS_KEY, async () =>
    persistRankings(HOUSE_RANKINGS_KEY, await computeHouseRankings(apiKey))
  )
}

export async function refreshSenateRankings(
  apiKey = resolveCongressApiKey()
): Promise<RankingsPayload<SenateMember>> {
  return singleFlight(SENATE_RANKINGS_KEY, async () =>
    persistRankings(SENATE_RANKINGS_KEY, await computeSenateRankings(apiKey))
  )
}

type TaggedMember =
  | { chamber: "house"; member: HouseMember }
  | { chamber: "senate"; member: SenateMember }

function interleave(house: HouseMember[], senate: SenateMember[]): TaggedMember[] {
  const out: TaggedMember[] = []
  const max = Math.max(house.length, senate.length)
  for (let i = 0; i < max; i++) {
    if (i < house.length) out.push({ chamber: "house", member: house[i] })
    if (i < senate.length) out.push({ chamber: "senate", member: senate[i] })
  }
  return out
}
export async function refreshAllRankings(
  apiKey = resolveCongressApiKey()
): Promise<{ house: RankingsPayload<HouseMember>; senate: RankingsPayload<SenateMember> }> {
  return singleFlight("all-rankings", async () => {
    const [houseMembers, senateMembers] = await Promise.all([
      fetchHouseMembers(apiKey),
      fetchSenateMembers(apiKey),
    ])

    const results = await mapWithConcurrency(
      interleave(houseMembers, senateMembers),
      FANOUT_CONCURRENCY,
      async (tagged) => {
        if (tagged.chamber === "house") {
          return { chamber: "house" as const, row: await getRankingRow(tagged.member) }
        }
        return { chamber: "senate" as const, row: await getRankingRow(tagged.member) }
      },
      FANOUT_DELAY_MS
    )

    const houseRows = results
      .filter(
        (r): r is { chamber: "house"; row: FanoutRow<HouseMember> } => r.chamber === "house"
      )
      .map((r) => r.row)
    const senateRows = results
      .filter(
        (r): r is { chamber: "senate"; row: FanoutRow<SenateMember> } => r.chamber === "senate"
      )
      .map((r) => r.row)
    const memberHoldings: MemberHoldings[] = [
      ...houseRows.map((row) => ({ row, chamber: "house" as const })),
      ...senateRows.map((row) => ({ row, chamber: "senate" as const })),
    ]
      .filter(({ row }) => row.ok)
      .map(({ row, chamber }) => ({
        bioguideId: row.id,
        memberName: row.name,
        party: row.party,
        chamber,
        positions: row.positions,
      }))
    writeBack(() => persistHoldings(memberHoldings))

    const [house, senate] = await Promise.all([
      persistRankings(HOUSE_RANKINGS_KEY, buildPayload(houseRows.map(stripFanoutExtras))),
      persistRankings(SENATE_RANKINGS_KEY, buildPayload(senateRows.map(stripFanoutExtras))),
    ])
    return { house, senate }
  })
}

export async function getHouseRankings(): Promise<RankingsPayload<HouseMember>> {
  const cached = await getCache<RankingsPayload<HouseMember>>(HOUSE_RANKINGS_KEY)
  if (cached && countPopulated(cached.byNetWorth) > 0) return cached
  const members = await fetchHouseMembers(resolveCongressApiKey())
  const fromDb = await rankingsFromDb(members)
  if (fromDb && countPopulated(fromDb.byNetWorth) > 0) {
    await setCache(HOUSE_RANKINGS_KEY, fromDb, RANKINGS_TTL_SECONDS)
    return fromDb
  }
  return membersToEmptyPayload(members)
}

export async function getSenateRankings(): Promise<RankingsPayload<SenateMember>> {
  const cached = await getCache<RankingsPayload<SenateMember>>(SENATE_RANKINGS_KEY)
  if (cached && countPopulated(cached.byNetWorth) > 0) return cached

  const members = await fetchSenateMembers(resolveCongressApiKey())

  const fromDb = await rankingsFromDb(members)
  if (fromDb && countPopulated(fromDb.byNetWorth) > 0) {
    await setCache(SENATE_RANKINGS_KEY, fromDb, RANKINGS_TTL_SECONDS)
    return fromDb
  }
  return membersToEmptyPayload(members)
}

File: backend/app/services/rankings.py
"""Net-worth / stock-holdings rankings (port of lib/rankings.ts).

Ranking rows are member dicts extended with `netWorth` and `stockHoldings`;
payloads are {"byNetWorth", "byStockHoldings", "generatedAt"} — identical keys
to the TS code so Redis cache entries stay interchangeable.
"""

import asyncio
import json
import logging
import math

from ..clients.congress import fetch_house_members, fetch_senate_members
from ..clients.quiver import QuiverCircuitOpenError, fetch_quiver_with_retry
from ..config import require_congress_api_key
from ..core.cache import get_cache, set_cache
from ..core.db import get_portfolios_from_db, upsert_portfolios, write_back
from ..core.util import map_with_concurrency, now_iso, single_flight
from .disclosures import get_disclosure_net_worth
from .stock_leaderboard import persist_holdings

logger = logging.getLogger("rankings")

HOUSE_RANKINGS_KEY = "house-rankings"
SENATE_RANKINGS_KEY = "senate-rankings"
RANKINGS_TTL_SECONDS = 2 * 60 * 60
FANOUT_CONCURRENCY = 1
FANOUT_DELAY_MS = 1500
CIRCUIT_OPEN_PAUSE_SECONDS = 65
MAX_CIRCUIT_WAITS = 3

QUIVER_HEADERS = {
    "Content-Type": "application/json",
    "User-Agent": "CongressMoneyMan/1.0",
    "X-Requested-With": "XMLHttpRequest",
}


def _is_finite_number(value) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool) and math.isfinite(value)


def _parse_json_array(value: str | None) -> list:
    if not value:
        return []
    try:
        parsed = json.loads(value)
    except (TypeError, json.JSONDecodeError):
        return []
    return parsed if isinstance(parsed, list) else []


def _get_live_net_worth(payload: dict) -> float | None:
    values = _parse_json_array(
        (payload.get("holdings_data") or {}).get("politician_net_worth_live")
    )
    value = values[0] if values else None
    return value if _is_finite_number(value) else None


def _get_live_stock_holdings(payload: dict) -> float | None:
    positions = _parse_json_array(
        (payload.get("live_stock_portfolio") or {}).get("live_stock_portfolio")
    )
    total = 0.0
    found_position = False
    for position in positions:
        if not isinstance(position, list) or len(position) < 2:
            continue
        value = position[1]
        if not _is_finite_number(value):
            continue
        total += value
        found_position = True
    return total if found_position else None


def _get_live_positions(payload: dict) -> list[dict]:
    positions = _parse_json_array(
        (payload.get("live_stock_portfolio") or {}).get("live_stock_portfolio")
    )
    out: list[dict] = []
    for position in positions:
        if not isinstance(position, list) or len(position) < 2:
            continue
        value = position[1]
        if not _is_finite_number(value) or value <= 0:
            continue
        symbol = position[0].strip().upper() if isinstance(position[0], str) else ""
        if not symbol:
            continue
        out.append({"ticker": symbol, "value": value})
    return out


def _ranking_sort_key(field: str):
    def key(row: dict):
        value = row.get(field)
        return (value is None, -(value or 0), row["name"])

    return key


async def _get_ranking_row(member: dict) -> dict:
    waits = 0
    while True:
        try:
            response = await fetch_quiver_with_retry(
                f"https://www.quiverquant.com/get_politician_page_tab_data/{member['id']}",
                QUIVER_HEADERS,
            )

            if response.status_code >= 400:
                logger.warning(
                    "%s tab-data HTTP %s", member["id"], response.status_code
                )
                return {
                    **member,
                    "netWorth": None,
                    "stockHoldings": None,
                    "positions": [],
                    "ok": False,
                }

            payload = response.json()
            return {
                **member,
                "netWorth": _get_live_net_worth(payload),
                "stockHoldings": _get_live_stock_holdings(payload),
                "positions": _get_live_positions(payload),
                "ok": True,
            }
        except QuiverCircuitOpenError:
            if waits < MAX_CIRCUIT_WAITS:
                waits += 1
                logger.warning(
                    "circuit open — pausing %ss before retrying %s",
                    CIRCUIT_OPEN_PAUSE_SECONDS,
                    member["id"],
                )
                await asyncio.sleep(CIRCUIT_OPEN_PAUSE_SECONDS)
                continue
            logger.warning("%s fan-out failed: circuit open", member["id"])
            return {
                **member,
                "netWorth": None,
                "stockHoldings": None,
                "positions": [],
                "ok": False,
            }
        except Exception as error:
            logger.warning("%s fan-out failed: %s", member["id"], error)
            return {
                **member,
                "netWorth": None,
                "stockHoldings": None,
                "positions": [],
                "ok": False,
            }


def _strip_fanout_extras(row: dict) -> dict:
    return {k: v for k, v in row.items() if k not in ("positions", "ok")}


def _build_payload(rows: list[dict]) -> dict:
    return {
        "byNetWorth": sorted(rows, key=_ranking_sort_key("netWorth")),
        "byStockHoldings": sorted(rows, key=_ranking_sort_key("stockHoldings")),
        "generatedAt": now_iso(),
    }


async def _apply_disclosure_fallback(payload: dict) -> dict:
    """Fill net-worth gaps from annual financial-disclosure estimates.

    Quiver's live figure stays authoritative — the FD estimate only fills rows
    where Quiver has none. Filled rows carry `netWorthSource: "fd"` and an
    `netWorthAsOf` year so the UI can flag them as annual-disclosure estimates
    rather than live figures. Members with a live figure get
    `netWorthSource: "quiver"`. Stock holdings are untouched: an FD lists assets
    in coarse ranges without live market values, so it can't produce the
    per-ticker live portfolio the holdings column shows."""
    fd_map = await get_disclosure_net_worth()
    if not fd_map:
        return payload

    rows = payload["byNetWorth"]
    filled = 0
    updated: list[dict] = []
    for row in rows:
        if row.get("netWorth") is not None:
            updated.append({**row, "netWorthSource": "quiver"})
            continue
        estimate = fd_map.get(row["id"])
        if not estimate:
            updated.append(row)
            continue
        filled += 1
        updated.append(
            {
                **row,
                "netWorth": estimate["netWorth"],
                "netWorthSource": "fd",
                "netWorthAsOf": estimate.get("asOf"),
            }
        )

    if filled == 0:
        return _build_payload(updated)

    logger.info("disclosure fallback filled %s net-worth gaps", filled)
    return _build_payload(updated)


def _count_populated(rows: list[dict]) -> int:
    return sum(
        1
        for row in rows
        if row.get("netWorth") is not None or row.get("stockHoldings") is not None
    )


def _members_to_empty_payload(members: list[dict]) -> dict:
    return _build_payload(
        [{**member, "netWorth": None, "stockHoldings": None} for member in members]
    )


def _merge_with_previous(fresh: dict, previous: dict | None) -> dict:
    if not previous:
        return fresh

    previous_by_id = {row["id"]: row for row in previous["byNetWorth"]}
    merged_rows = []
    for row in fresh["byNetWorth"]:
        prior = previous_by_id.get(row["id"])
        if not prior:
            merged_rows.append(row)
            continue
        merged_rows.append(
            {
                **row,
                "netWorth": row["netWorth"] if row["netWorth"] is not None else prior.get("netWorth"),
                "stockHoldings": row["stockHoldings"]
                if row["stockHoldings"] is not None
                else prior.get("stockHoldings"),
            }
        )

    return _build_payload(merged_rows)


async def _persist_rankings(key: str, fresh: dict) -> dict:
    previous = await get_cache(key)
    merged = _merge_with_previous(fresh, previous)

    await set_cache(key, merged, RANKINGS_TTL_SECONDS)
    logger.info(
        "cached %s: %s/%s members populated",
        key,
        _count_populated(merged["byNetWorth"]),
        len(merged["byNetWorth"]),
    )
    portfolio_rows = [
        {
            "bioguide_id": row["id"],
            "net_worth": row["netWorth"],
            "stock_holdings": row["stockHoldings"],
        }
        for row in merged["byNetWorth"]
        if row.get("netWorth") is not None or row.get("stockHoldings") is not None
    ]
    write_back(upsert_portfolios(portfolio_rows))

    return merged


async def _rankings_from_db(members: list[dict]) -> dict | None:
    portfolios = await get_portfolios_from_db()
    if not portfolios:
        return None

    by_id = {p["bioguide_id"]: p for p in portfolios}
    rows = []
    for member in members:
        portfolio = by_id.get(member["id"]) or {}
        rows.append(
            {
                **member,
                "netWorth": portfolio.get("net_worth"),
                "stockHoldings": portfolio.get("stock_holdings"),
            }
        )
    return _build_payload(rows)


async def compute_house_rankings(api_key: str | None = None) -> dict:
    api_key = api_key or require_congress_api_key()
    members = await fetch_house_members(api_key)
    rows = await map_with_concurrency(
        members, FANOUT_CONCURRENCY, _get_ranking_row, FANOUT_DELAY_MS
    )
    return _build_payload([_strip_fanout_extras(row) for row in rows])


async def compute_senate_rankings(api_key: str | None = None) -> dict:
    api_key = api_key or require_congress_api_key()
    members = await fetch_senate_members(api_key)
    rows = await map_with_concurrency(
        members, FANOUT_CONCURRENCY, _get_ranking_row, FANOUT_DELAY_MS
    )
    return _build_payload([_strip_fanout_extras(row) for row in rows])


async def refresh_house_rankings(api_key: str | None = None) -> dict:
    api_key = api_key or require_congress_api_key()

    async def run() -> dict:
        return await _persist_rankings(
            HOUSE_RANKINGS_KEY, await compute_house_rankings(api_key)
        )

    return await single_flight(HOUSE_RANKINGS_KEY, run)


async def refresh_senate_rankings(api_key: str | None = None) -> dict:
    api_key = api_key or require_congress_api_key()

    async def run() -> dict:
        return await _persist_rankings(
            SENATE_RANKINGS_KEY, await compute_senate_rankings(api_key)
        )

    return await single_flight(SENATE_RANKINGS_KEY, run)


def _interleave(house: list[dict], senate: list[dict]) -> list[dict]:
    out: list[dict] = []
    for i in range(max(len(house), len(senate))):
        if i < len(house):
            out.append({"chamber": "house", "member": house[i]})
        if i < len(senate):
            out.append({"chamber": "senate", "member": senate[i]})
    return out


async def refresh_all_rankings(api_key: str | None = None) -> dict:
    api_key = api_key or require_congress_api_key()

    async def run() -> dict:
        house_members, senate_members = await asyncio.gather(
            fetch_house_members(api_key), fetch_senate_members(api_key)
        )

        async def fan_out(tagged: dict) -> dict:
            return {
                "chamber": tagged["chamber"],
                "row": await _get_ranking_row(tagged["member"]),
            }

        results = await map_with_concurrency(
            _interleave(house_members, senate_members),
            FANOUT_CONCURRENCY,
            fan_out,
            FANOUT_DELAY_MS,
        )

        house_rows = [r["row"] for r in results if r["chamber"] == "house"]
        senate_rows = [r["row"] for r in results if r["chamber"] == "senate"]

        member_holdings = [
            {
                "bioguideId": row["id"],
                "memberName": row["name"],
                "party": row["party"],
                "chamber": chamber,
                "positions": row["positions"],
            }
            for rows, chamber in ((house_rows, "house"), (senate_rows, "senate"))
            for row in rows
            if row["ok"]
        ]
        write_back(persist_holdings(member_holdings))

        house, senate = await asyncio.gather(
            _persist_rankings(
                HOUSE_RANKINGS_KEY,
                _build_payload([_strip_fanout_extras(r) for r in house_rows]),
            ),
            _persist_rankings(
                SENATE_RANKINGS_KEY,
                _build_payload([_strip_fanout_extras(r) for r in senate_rows]),
            ),
        )
        return {"house": house, "senate": senate}

    return await single_flight("all-rankings", run)


async def get_house_rankings() -> dict:
    cached = await get_cache(HOUSE_RANKINGS_KEY)
    if cached and _count_populated(cached["byNetWorth"]) > 0:
        return await _apply_disclosure_fallback(cached)
    members = await fetch_house_members(require_congress_api_key())
    from_db = await _rankings_from_db(members)
    if from_db and _count_populated(from_db["byNetWorth"]) > 0:
        await set_cache(HOUSE_RANKINGS_KEY, from_db, RANKINGS_TTL_SECONDS)
        return await _apply_disclosure_fallback(from_db)
    return await _apply_disclosure_fallback(_members_to_empty_payload(members))


async def get_senate_rankings() -> dict:
    cached = await get_cache(SENATE_RANKINGS_KEY)
    if cached and _count_populated(cached["byNetWorth"]) > 0:
        return await _apply_disclosure_fallback(cached)

    members = await fetch_senate_members(require_congress_api_key())

    from_db = await _rankings_from_db(members)
    if from_db and _count_populated(from_db["byNetWorth"]) > 0:
        await set_cache(SENATE_RANKINGS_KEY, from_db, RANKINGS_TTL_SECONDS)
        return await _apply_disclosure_fallback(from_db)
    return await _apply_disclosure_fallback(_members_to_empty_payload(members))
